### ISSUE TITLE
Need info on achieving horizontal autoscaling for Job pods

### ISSUE THREAD

**Issue by kmathewmk** [kubernetes/website#41394]
Could you please describe how horizontal autoscaling can be acheived for Job pods?

**Comment by dipesh-rawat** [comment]
/retitle Need info on achieving horizontal autoscaling for Job pods

**Comment by dipesh-rawat** [comment]
@kmathewmk I would recommend consider exploring [Server Fault](https://serverfault.com/questions/tagged/kubernetes) and posting your query there if a similar question hasn't already been addressed. 

Additionally, you can leverage the [Kubernetes Slack](http://slack.k8s.io/) or the [Discuss Kubernetes](https://discuss.kubernetes.io/) forum to share your question and seek assistance from the community.

**Comment by niranjandarshann** [comment]
Horizontal autoscaling for Job pods in Kubernetes can be achieved by utilizing the Horizontal Pod Autoscaler (HPA) .
Default behavior of Kubernetes autoscaling is not designed for  pods job since jobs are meant to run to completion rather than scale dynamically.
 
you can follow these steps:

1.  Deploy your Job pod: using a Kubernetes manifest or deployment file.

 2.  Set up the Horizontal Pod Autoscaler (HPA): Create an HPA resource that targets the Job pods with minimum and maximum number of replicas we  want to have.
  3.  Configure custom metrics (we can or cannot ): If we want to use custom metrics for autoscaling, we need to set up a custom metrics server and expose the metric we want to use.

  4.  Update the HPA configuration: Modify the HPA configuration to use the custom metric instead of the default CPU utilization metric Specify the custom metric name and the target value or threshold for scaling.

5.    Apply the HPA configuration: Apply the updated HPA configuration to our Kubernetes cluster using the `kubectl apply` command This will create or update the HPA resource.

  6.   Monitor and observe: Kubernetes will now monitor the metric specified in the HPA configuration and adjust the number of replicas for the Job pod accordingly we can use the `kubectl get hpa` command to view the current status of the HPA and monitor the scaling behavior.

I think these points help you in understanding solution.

**Comment by kmathewmk** [comment]
After some searching, there are 3 kinds of autoscaling in Kubernetes:

1. Horizontal Pod Autoscaler
2. Vertical Pod Autoscaler
3. Cluster Autoscaler

[Horizontal Pod Autoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale) is supported only for Deployment, ReplicaSet, StatefulSet, or replication controller by name 
[Vertical Pod Autoscaler](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/README.md) automatically updates the "requests" settings of a pod based on usage and allowing scheduling onto appropriate nodes
[Cluster Autoscaler](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/README.md) is cloud provider specific. Its spins nodes up and down depending on the resources requested by active pods.

By setting the "cpu" and "memory" under "resources.requests" setting for the Job pod spec, I was able to get Job pods to trigger autoscaling of nodes. When existing nodes could not accomodate the resource requests of pods, new nodes were spun up and when the jobs were completed the nodes were spun down.

It would be good if [Kubernetes doc](https://kubernetes.io/docs/tasks/run-application/) covered all the 3 major types of autoscaling

**Comment by sftim** [comment]
The https://kubernetes.io/docs/tasks/run-application/ task list isn't about _any_ kind of autoscaling. It's about running applications. If anything, we should move the autoscaling pages out of their task list into another section such as “Configure automatic scaling”.

@kmathewmk if that section existed, would it (or could it) cover the need you've identified?

**Comment by kmathewmk** [comment]
Yes, auto scaling could possible have a section of its own for eg under "Concepts".

**Comment by sftim** [comment]
> Yes, auto scaling could possible have a section of its own for eg under "Concepts".

See https://github.com/kubernetes/website/issues/4493#ref-issue-2010745978
